### PR TITLE
Damage tracking for view children

### PR DESCRIPTION
**Pull request by emersion**
Test plan: make sure damage is accumulated when view children commit.

* Add `wlr_log(L_DEBUG, "child damaged");` in `view.c` on line 231
* Add `wlr_log(L_DEBUG, "unmanaged surface damaged");` in `xwayland.c` on line 32

Check that these appear in the log when opening a popup and a combo in `gnome-calculator`, under xdg-shell and xwayland.

### CODE REVIEW

**Comment by emersion** [comment]
This PR is ready for review.

**Comment by SirCmpwn** [comment]
Can you add a utility function to this PR for me? `container_damage_whole` which just damages the container's whole box on its output.

File: include/sway/tree/view.h [hunk 1]
```diff
@@ -95,6 +106,37 @@ struct sway_wl_shell_view {
...
 };
 
+struct sway_view_child;
+
+struct sway_view_child_impl {
+	void (*destroy)(struct sway_view_child *child);
+};
+
+/**
+ * A view child is a surface in the view tree, such as a subsurface or a popup.
```
Comment (SirCmpwn) [inline include/sway/tree/view.h]: I'm so happy that popups are children of their parent windows now

Screw you wlc

File: sway/desktop/xwayland.c [hunk 2]
```diff
@@ -2,42 +2,94 @@
...
+	wlr_xwayland_surface_configure(xsurface, ev->x, ev->y,
+		ev->width, ev->height);
+}
+
+static void unmanaged_handle_commit(struct wl_listener *listener, void *data) {
+	struct sway_xwayland_unmanaged *surface =
+		wl_container_of(listener, surface, commit);
+	struct wlr_xwayland_surface *xsurface = surface->wlr_xwayland_surface;
+	desktop_damage_from_surface(xsurface->surface, xsurface->x, xsurface->y);
+	// TODO: handle window motion
```
Comment (SirCmpwn) [inline sway/desktop/xwayland.c]: :<

Comment (emersion) [reply sway/desktop/xwayland.c]: This reason it's a TODO is that nobody moves an unmanaged window, so it's difficult to test.

File: sway/tree/view.c [hunk 3]
```diff
@@ -136,22 +151,33 @@ void view_map(struct sway_view *view, struct wlr_surface *wlr_surface) {
...
+		&view->surface_new_subsurface);
+	view->surface_new_subsurface.notify = view_handle_surface_new_subsurface;
+
 	arrange_windows(cont->parent, -1, -1);
 	input_manager_set_focus(input_manager, cont);
 
 	view_damage_whole(view);
 	view_update_outputs(view, NULL);
+
+	// TODO: create view children for subsurfaces
```
Comment (SirCmpwn) [inline sway/tree/view.c]: :<

File: sway/desktop/xwayland.c (hunk 2 above)
Comment (SirCmpwn) [reply sway/desktop/xwayland.c]: Can we at least do an implementation that might work, then?

File: sway/tree/container.c [hunk 4]
```diff
@@ -512,9 +512,18 @@ static bool find_child_func(struct sway_container *con, void *data) {
...
 }
+
+void container_damage_whole(struct sway_container *con) {
+	struct sway_container *output = con;
+	if (output->type != C_OUTPUT) {
+		output = container_parent(output, C_OUTPUT);
+	}
+
+	output_damage_whole_rect(output->sway_output, con->x, con->y, con->width,
+		con->height);
```
Comment (SirCmpwn) [inline sway/tree/container.c]: subsurfaces tho

File: sway/tree/view.c (hunk 3 above)
Comment (emersion) [reply sway/tree/view.c]: Removed that outdated TODO. It's fixed by `view_init_subsurfaces` a few lines above.

File: sway/tree/container.c (hunk 4 above)
Comment (emersion) [reply sway/tree/container.c]: Err, you're right, I can't do this

File: sway/desktop/xwayland.c (hunk 2 above)
Comment (emersion) [reply sway/desktop/xwayland.c]: ack

**Pull request merged**